python/ray/data/tests/test_clickhouse.py

```diff
@@ -0,0 +1,130 @@
+from unittest import mock
+from unittest.mock import MagicMock
+
+import pyarrow as pa
+import pytest
+
+from ray.data._internal.datasource.clickhouse_datasource import ClickHouseDatasource
+
+
+@pytest.fixture
+def mock_clickhouse_client():
+    client_mock = mock.MagicMock()
+    client_mock.return_value = client_mock
+    return client_mock
+
+
+class TestClickHouseDatasource:
+    """Tests for ClickHouseDatasource."""
+
+    @pytest.fixture
+    def datasource(self, mock_clickhouse_client):
+        datasource = ClickHouseDatasource(
+            table="default.table_name",
+            dsn="clickhouse://user:password@localhost:8123/default",
+            columns=["column1", "column2"],
+            order_by=(["column1"], False),
+            client_settings={"setting1": "value1"},
+            client_kwargs={"arg1": "value1"},
+        )
+        datasource._client = mock_clickhouse_client
+        return datasource
+
+    def test_init(self, datasource):
+        # Test query generation with columns, filters, and order_by
+        expected_query = (
+            "SELECT column1, column2 FROM default.table_name " "ORDER BY column1"
+        )
+        assert datasource._query == expected_query
+
+    def test_estimate_inmemory_data_size(self, datasource):
+        mock_client = mock.MagicMock()
+        datasource._init_client = MagicMock(return_value=mock_client)
+        mock_client.query.return_value.result_rows = [[12345]]
+        size = datasource.estimate_inmemory_data_size()
+        assert size == 12345
+        mock_client.query.assert_called_once_with(
+            f"SELECT SUM(byteSize(*)) AS estimate FROM ({datasource._query})"
+        )
+
+    @pytest.mark.parametrize(
+        "columns, expected_query_part",
+        [
+            (
+                [
+                    "field1",
+                ],
+                "SELECT field1 FROM default.table_name",
+            ),
+            (["field1", "field2"], "SELECT field1, field2 FROM default.table_name"),
+            (None, "SELECT * FROM default.table_name"),
+        ],
+    )
+    def test_generate_query_columns(self, datasource, columns, expected_query_part):
```

Comment:
Can we please also add a test generating the full query (so that we certain e2e flow works as expected)

Reply:
Added!